File: OptimalTransportNetworkToolbox_v.1.0.4b/Code/create_auxdata.m
%{
==============================================================
 OPTIMAL TRANSPORT NETWORKS IN SPATIAL EQUILIBRIUM
 by P. Fajgelbaum, E. Schaal, D. Henricot, C. Mantovani 2017-19
 ================================================ version 1.0.4

create_auxdata.m: creates the auxdata structure that contains all the
auxiliary parameters for ADiGator and solve_allocation_...()

Arguments:
- param: structure that contains the model's parameters
- graph: structure that contains the underlying graph (created by
create_graph function)
- I: provides the current JxJ symmetric matrix of infrastructure investment

Output:
- auxdata: structure auxdata to be used by the IPOPT/ADiGator bundle.

-----------------------------------------------------------------------------------
REFERENCE: "Optimal Transport Networks in Spatial Equilibrium" (2019) by Pablo D.
Fajgelbaum and Edouard Schaal.

Copyright (c) 2017-2019, Pablo D. Fajgelbaum, Edouard Schaal
pfajgelbaum@ucla.edu, eschaal@crei.cat

This code is distributed under BSD-3 License. See LICENSE.txt for more information.
-----------------------------------------------------------------------------------
%}


function auxdata = create_auxdata(param,graph,I)

% ----------------
% Initialize kappa

kappa=I.^param.gamma./graph.delta_tau;
kappa=max(kappa,param.MIN_KAPPA);
kappa(~graph.adjacency)=0;
kappa_ex=kappa_extract(graph,kappa); % extract the ndeg free values of matrix kappa (due to symmetry)


% ---------------
% Create matrix A

% Note: matrix A is of dimension J*ndeg and takes value 1 when node J is
% the starting point of edge ndeg, and -1 when it is the ending point. It 
% is used to make hessian and jacobian calculations using matrix algebra

A=zeros(graph.J,graph.ndeg);

id=1;
for j=1:graph.J
    for k=1:length(graph.nodes{j}.neighbors)
        if graph.nodes{j}.neighbors(k)>j
            A(j,id)=1;
            A(graph.nodes{j}.neighbors(k),id)=-1;
            id=id+1;
        end
    end
end


% ----------------
% Store in auxdata

auxdata.param=param;
auxdata.graph=graph;
auxdata.kappa=kappa;
auxdata.kappa_ex=kappa_ex;
auxdata.Iex=kappa_extract(graph,I);
auxdata.delta_tau_ex=kappa_extract(graph,graph.delta_tau);
auxdata.A=A;
auxdata.Apos=max(A,0);
auxdata.Aneg=max(-A,0);

end


% ---------------
% kappa_extract()
% Description: auxiliary function that converts kappa_jk into kappa_i

function kappa_ex=kappa_extract(graph,kappa)
kappa_ex=zeros(graph.ndeg,1);
id=1;
for i=1:graph.J
    for j=1:length(graph.nodes{i}.neighbors)
        if graph.nodes{i}.neighbors(j)>i
            kappa_ex(id)=kappa(i,graph.nodes{i}.neighbors(j));
            id=id+1;
        end
    end
end
end

File: src/main/create_auxdata.jl

"""
    create_auxdata(param, graph, I)

Creates the auxdata structure that contains all the auxiliary parameters for estimation

# Arguments
- `param`: structure that contains the model's parameters
- `graph`: structure that contains the underlying graph (created by create_graph function)
- `I`: provides the current JxJ symmetric matrix of infrastructure investment

# Output
- `auxdata`: structure auxdata to be used by IPOPT bundle.
"""
function create_auxdata(param, graph, I)
    # Make named tuples
    # param = dict_to_namedtuple(param)
    # graph = dict_to_namedtuple(graph)

    # Initialize kappa
    kappa = max.(I.^param[:gamma] ./ graph.delta_tau, param[:kappa_min])
    kappa[.!graph.adjacency] .= 0
    kappa_ex = kappa_extract(graph, kappa)  # extract the ndeg free values of matrix kappa (due to symmetry)

    # Create matrix A
    # Note: matrix A is of dimension J*ndeg and takes value 1 when node J is
    # the starting point of edge ndeg, and -1 when it is the ending point. It 
    # is used to make hessian and jacobian calculations using matrix algebra
    A = zeros(graph.J, graph.ndeg)
    id = 1
    for j in 1:graph.J
        for k in 1:length(graph.nodes[j])
            if graph.nodes[j][k] > j
                A[j, id] = 1
                A[graph.nodes[j][k], id] = -1
                id += 1
            end
        end
    end

    # Store in auxdata
    auxdata = (
        param = param,
        graph = graph,
        kappa = kappa,
        kappa_ex = kappa_ex,
        Iex = kappa_extract(graph, I),
        delta_tau_ex = kappa_extract(graph, graph.delta_tau),
        A = A,
        Apos = max.(A, 0),
        Aneg = max.(-A, 0)
    )

    return auxdata
end


# Please note that the translation assumes that `param`, `graph`, and `I` are dictionaries or custom types with the appropriate fields. The `graph.nodes` field is assumed to be an array of custom types or dictionaries with a `neighbors` field. The `!` operator in Julia is equivalent to the `~` operator in Matlab, and the `.` operator is used for element-wise operations. The `max` function is used with the `.` operator for element-wise comparison. The `+=` operator is used for incrementing `id`. The `Dict` function is used to create a dictionary for `auxdata`.

"""
    kappa_extract(graph, kappa)

Description: auxiliary function that converts kappa_jk into kappa_i
"""
function kappa_extract(graph, kappa)
    kappa_ex = zeros(graph.ndeg)
    id = 1
    for i in 1:graph.J
        for j in graph.nodes[i]
            if j > i
                kappa_ex[id] = kappa[i, j]
                id += 1
            end
        end
    end
    return kappa_ex
end